pkg/syncer/k8scloudoperator/placement.go

```diff
@@ -410,6 +421,221 @@ func stampPVCWithError(ctx context.Context, client kubernetes.Interface,
+	}
+
+	for _, pvcItem := range allPVCs.Items {
+		allPVCsMap[pvcItem.Name] = pvcItem
+	}
+	return allPVCsMap, nil
+}
+
+// Check if internal FSS to check for sibling replica bound PVCs is enabled
+func isSiblingReplicaBoundPvcFSSEnabled(ctx context.Context) bool {
+	log := logger.GetLogger(ctx)
+
+	containerOrchestratorUtility := commonco.ContainerOrchestratorUtility
+	if containerOrchestratorUtility == nil {
+		clusterFlavor, err := cnsconfig.GetClusterFlavor(ctx)
+		if err != nil {
+			log.Debugf("Failed retrieving cluster flavor. Error: %v", err)
+			return false
+		}
+		containerOrchestratorUtility, err = commonco.GetContainerOrchestratorInterface(ctx, common.Kubernetes, clusterFlavor, *admissionhandler.COInitParams)
+		if err != nil {
+			log.Debugf("failed to get k8s interface. err: %v", err)
+			return false
+		}
+	}
+
+	return containerOrchestratorUtility.IsFSSEnabled(ctx, common.SiblingReplicaBoundPvcCheck)
+}
+
+// eliminateNodesWithPvcOfSiblingReplica filters out the nodes that have bound PVCs of sibling replicas.
+// It finds cousin PVCs of currPVC and elimates all nodes which have at least one cousin PVC placed on them.
+func eliminateNodesWithPvcOfSiblingReplica(ctx context.Context, client kubernetes.Interface, currPVC *v1.PersistentVolumeClaim, candidateHosts []string) ([]string, error) {
+	log := logger.GetLogger(ctx)
+
+	if !isSiblingReplicaBoundPvcFSSEnabled(ctx) {
+		log.Infof("FSS to check for sibling replica's PVCs is not enabled.")
+		return candidateHosts, nil
+	}
+
+	// Proceed only if it is a vDPP PVC
+	pvcLabels := currPVC.GetLabels()
+	appplatformLabelVal, labelOk := pvcLabels[appplatformLabel]
+	if !labelOk {
+		log.Infof("Not a vDPP PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	// Do not proceed if it is a standalone pod
+	pvcAnnotation := currPVC.ObjectMeta.Annotations
+	if _, ok := pvcAnnotation[nodeAffinityAnnotationKey]; ok {
+		log.Infof("Node affinity key already specified for PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	if val, ok := pvcLabels[siblingReplicaNodeSnatchingCheckOptOutLabel]; ok && val == "true" {
+		// siblingReplicaNodeSnatchingCheckOptOut is opted in by default unless specified otherwise
+		log.Infof("Sibling Replica Bound PVC check is not opted, skip this step for PVC %s.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	cousinPVCs, err := getCousinPVCs(ctx, client, currPVC)
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	selectLabel := fmt.Sprintf("%s=%s", appplatformLabel, appplatformLabelVal)
+	allPVCs, err := getAllPvcs(ctx, client, currPVC.Namespace, selectLabel)
+	if err != nil {
+		return candidateHosts, err
+	}
+
+	for _, cousinPVC := range cousinPVCs {
+		pvc, ok := allPVCs[cousinPVC]
+		if !ok {
+			// This scenario can happen if psp-operator PVC watcher has already deleted the PVC
+			log.Errorf("Failed to find PVC object for PVC %s", cousinPVC)
+			continue
+		}
+
+		pvName := pvc.Spec.VolumeName
```

Comment:
It is possible for some of the PVCs, volume may not be created yet and `pvc.Spec.VolumeName` could be nil, we should ignore such PVCs.